File: EcSignatureConverter/EcSignatureConverterDerToP1363.java
package EC_Signature_secp256r1_conversion;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.Base64;

public class EcSignatureConverterDerToP1363 {
    public static void main(String[] args) throws Exception {
        System.out.println("EC signature converter DER to P1363 encoding");
        System.out.println("Please note that all values are in Base64 encoded form");

        // insert the ecdsa signature in DER encoding
        String signatureDerBase64 = "MEUCIQCW98vDD+c/Wx3WX3T+Ph9ZDf1s5nuDThz2xrhSOjwGFAIgNpUg5SygNK7W+zzw4eNZkivizpU/UUXMMR2zPw2D7J4=";

        String signatureP1363Base64 = convertSignatureDerToP1363Base64(signatureDerBase64);
        System.out.println("signature in DER encoding:   " + signatureDerBase64);
        System.out.println("signature in P1363 encoding: " + signatureP1363Base64);
    }

    // convertPlainToDer and convertDerToPlain
    // source https://stackoverflow.com/a/61873962/8166854  answered May 18 at 16:07 dave_thompson_085
    // secp384r1 (aka P-384) has 384-bit order so use 384/8 which is 48 for n
    private static String convertSignatureP1363ToDerBase64 (String plainBase64) {
        byte[] plain = base64Decoding(plainBase64);
        int n = 32; // for example assume 256-bit-order curve like P-256
        BigInteger r = new BigInteger (+1, Arrays.copyOfRange(plain,0,n));
        BigInteger s = new BigInteger (+1, Arrays.copyOfRange(plain,n,n*2));
        byte[] x1 = r.toByteArray(), x2 = s.toByteArray();
        // already trimmed two's complement, as DER wants
        int len = x1.length + x2.length + (2+2), idx = len>=128? 3: 2;
        // the len>=128 case can only occur for curves of 488 bits or more,
        // and can be removed if you will definitely not use such curve(s)
        byte[] out = new byte[idx+len];
        out[0] = 0x30;
        if( idx==3 ){
            out[1] = (byte)0x81;
            out[2] = (byte)len; }
        else {
            out[1] = (byte)len; }
        out[idx] = 2;
        out[idx+1] = (byte)x1.length;
        System.arraycopy(x1, 0, out, idx+2, x1.length);
        idx += x1.length + 2;
        out[idx] = 2;
        out[idx+1] = (byte)x2.length;
        System.arraycopy(x2, 0, out, idx+2, x2.length);
        return base64Encoding(out);
    }

    private static String convertSignatureDerToP1363Base64 (String derBase64) throws Exception {
        byte[] der = base64Decoding(derBase64);
        int n = 32; // for example assume 256-bit-order curve like P-256
        BigInteger r, s;
        byte[] out;
        if( der[0] != 0x30 ) throw new Exception();
        int idx = der[1]==0x81? 3: 2; // the 0x81 case only occurs for curve over 488 bits
        if( der[idx] != 2 ) throw new Exception();
        r = new BigInteger (1, Arrays.copyOfRange(der,  idx+2, idx+2+der[idx+1]));
        idx += der[idx+1] + 2;
        if( der[idx] != 2 ) throw new Exception();
        s = new BigInteger (1, Arrays.copyOfRange(der,  idx+2, idx+2+der[idx+1]));
        if( idx + der[idx+1] + 2 != der.length ) throw new Exception();
        // common output
        out = new byte[2*n];
        toFixed(r, out, 0, n);
        toFixed(s, out, n, n);
        return base64Encoding(out);
    }

    static void toFixed (BigInteger x, byte[] a, int off, int len) throws Exception {
        byte[] t = x.toByteArray();
        if( t.length == len+1 && t[0] == 0 ) System.arraycopy (t,1, a,off, len);
        else if( t.length <= len ) System.arraycopy (t,0, a,off+len-t.length, t.length);
        else throw new Exception();
    }

    private static String base64Encoding(byte[] input) {
        return Base64.getEncoder().encodeToString(input);
    }
    private static byte[] base64Decoding(String input) {
        return Base64.getDecoder().decode(input);
    }
}

File: EcSignatureConverter/EcSignatureConverterDerToP1363.php
<?php

/*
 * ASN.1 <-> DER converter taken from https://github.com/firebase/php-jwt/blob/master/src/JWT.php
 * author   Neuman Vong <neuman@twilio.com>
 * author   Anant Narayanan <anant@php.net>
 * license  http://opensource.org/licenses/BSD-3-Clause 3-clause BSD
 */
const ASN1_INTEGER = 0x02;
const ASN1_SEQUENCE = 0x10;
const ASN1_BIT_STRING = 0x03;

/**
 * Convert an ECDSA signature to an ASN.1 DER sequence
 *
 * @param string $sig The ECDSA signature to convert
 * @return  string The encoded DER object
 */
function signatureToDER($sig)
{
    // Separate the signature into r-value and s-value
    list($r, $s) = \str_split($sig, (int)(\strlen($sig) / 2));

    // Trim leading zeros
    $r = \ltrim($r, "\x00");
    $s = \ltrim($s, "\x00");

    // Convert r-value and s-value from unsigned big-endian integers to
    // signed two's complement
    if (\ord($r[0]) > 0x7f) {
        $r = "\x00" . $r;
    }
    if (\ord($s[0]) > 0x7f) {
        $s = "\x00" . $s;
    }

    return encodeDER(
        ASN1_SEQUENCE,
        encodeDER(ASN1_INTEGER, $r) .
        encodeDER(ASN1_INTEGER, $s)
    );
}

/**
 * Encodes a value into a DER object.
 *
 * @param int $type DER tag
 * @param string $value the value to encode
 * @return  string  the encoded object
 */
function encodeDER($type, $value)
{
    $tag_header = 0;
    if ($type === ASN1_SEQUENCE) {
        $tag_header |= 0x20;
    }

    // Type
    $der = \chr($tag_header | $type);

    // Length
    $der .= \chr(\strlen($value));

    return $der . $value;
}

/**
 * Encodes signature from a DER object.
 *
 * @param string $der binary signature in DER format
 * @param int $keySize the number of bits in the key
 * @return  string  the signature
 */
function signatureFromDER($der, $keySize)
{
    // OpenSSL returns the ECDSA signatures as a binary ASN.1 DER SEQUENCE
    list($offset, $_) = readDER($der);
    list($offset, $r) = readDER($der, $offset);
    list($offset, $s) = readDER($der, $offset);

    // Convert r-value and s-value from signed two's compliment to unsigned
    // big-endian integers
    $r = \ltrim($r, "\x00");
    $s = \ltrim($s, "\x00");

    // Pad out r and s so that they are $keySize bits long
    $r = \str_pad($r, $keySize / 8, "\x00", STR_PAD_LEFT);
    $s = \str_pad($s, $keySize / 8, "\x00", STR_PAD_LEFT);

    return $r . $s;
}

/**
 * Reads binary DER-encoded data and decodes into a single object
 *
 * @param string $der the binary data in DER format
 * @param int $offset the offset of the data stream containing the object
 * to decode
 * @return array [$offset, $data] the new offset and the decoded object
 */
function readDER($der, $offset = 0)
{
    $pos = $offset;
    $size = \strlen($der);
    $constructed = (\ord($der[$pos]) >> 5) & 0x01;
    $type = \ord($der[$pos++]) & 0x1f;
    // Length
    $len = \ord($der[$pos++]);
    if ($len & 0x80) {
        $n = $len & 0x1f;
        $len = 0;
        while ($n-- && $pos < $size) {
            $len = ($len << 8) | \ord($der[$pos++]);
        }
    }
    // Value
    if ($type == ASN1_BIT_STRING) {
        $pos++; // Skip the first contents octet (padding indicator)
        $data = \substr($der, $pos, $len - 1);
        $pos += $len - 1;
    } elseif (!$constructed) {
        $data = \substr($der, $pos, $len);
        $pos += $len;
    } else {
        $data = null;
    }

    return array($pos, $data);
}

// own functions

function convertSignatureDerToP1363Base64($signatureDer)
{
    $signature = base64Decoding($signatureDer);
    return base64Encoding(signatureFromDER($signature, 256));
    $sigToJavaBase64 = base64Encoding($sigToJava);
}

function convertSignatureP1363ToDerBase64($signatureP1363)
{
    $signature = base64Decoding($signatureP1363);
    return base64Encoding(signatureToDER($signature));
}

function base64Encoding($input)
{
    return base64_encode($input);
}

function base64Decoding($input)
{
    return base64_decode($input);
}

echo 'EC signature converter DER to P1363 encoding' . PHP_EOL;
echo 'Please note that all values are in Base64 encoded form' . PHP_EOL;

// insert the ecdsa signature in DER encoding
$signatureDerBase64 = "MEUCIQCW98vDD+c/Wx3WX3T+Ph9ZDf1s5nuDThz2xrhSOjwGFAIgNpUg5SygNK7W+zzw4eNZkivizpU/UUXMMR2zPw2D7J4=";

$signatureP1363Base64 = convertSignatureDerToP1363Base64($signatureDerBase64);

echo 'signature in DER encoding:   ' . $signatureDerBase64 . PHP_EOL;
echo 'signature in P1363 encoding: ' . $signatureP1363Base64 . PHP_EOL;

?>
